### Linear Regression with Cross Validation

## Merge feature and test datasets

In [24]:
### Importing the required packages and libraries
# we will need numpy and pandas later
import numpy as np
import pandas as pd


from sklearn.linear_model import LinearRegression
#import DecisionTreeClassifier class
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
# Randomly splitting the original dataset into training set and testing set
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

In [5]:
# reading a CSV file directly from Web, and store it in a pandas DataFrame:
# "read_csv" is a pandas function to read csv files from web or local device:
walmart_features_df =  pd.read_csv('features.csv')

walmart_sampleSubmission_df =  pd.read_csv('sampleSubmission.csv')

walmart_stores_df =  pd.read_csv('stores.csv')

walmart_test_df =  pd.read_csv('test.csv')

walmart_train_df =  pd.read_csv('train.csv')

In [6]:
walmart_features_df.fillna(value=0, inplace=True)

#adding all the markdown sales and putting it in one column
walmart_features_df['Markdowns'] = walmart_features_df['MarkDown1'] + walmart_features_df['MarkDown2'] + walmart_features_df['MarkDown3'] + walmart_features_df['MarkDown4'] + walmart_features_df['MarkDown5'] 

labelsToDrop = ['MarkDown1', 'MarkDown2', 'MarkDown3','MarkDown4','MarkDown5']
walmart_features_df.drop(labels=labelsToDrop,axis=1, inplace=True)

In [7]:
#merging datasets, joining walmart_Store_Df to features by the common column of store.
MergeFeatureAndStore_Df = pd.merge(walmart_features_df,
                 walmart_stores_df[['Store','Size']],
                 on='Store')

MergeFeatureAndStore_Df.head()

Store        Date  Temperature  Fuel_Price         CPI  Unemployment  \
0      1  2010-02-05        42.31       2.572  211.096358         8.106   
1      1  2010-02-12        38.51       2.548  211.242170         8.106   
2      1  2010-02-19        39.93       2.514  211.289143         8.106   
3      1  2010-02-26        46.63       2.561  211.319643         8.106   
4      1  2010-03-05        46.50       2.625  211.350143         8.106   

   IsHoliday  Markdowns    Size  
0      False        0.0  151315  
1       True        0.0  151315  
2      False        0.0  151315  
3      False        0.0  151315  
4      False        0.0  151315

In [8]:
#merging train dataset
Merged_Train_Features_Store_Df = pd.merge(MergeFeatureAndStore_Df, walmart_train_df[['Date','Dept','Weekly_Sales']], on='Date')
Merged_Train_Features_Store_Df.head()

Store        Date  Temperature  Fuel_Price         CPI  Unemployment  \
0      1  2010-02-05        42.31       2.572  211.096358         8.106   
1      1  2010-02-05        42.31       2.572  211.096358         8.106   
2      1  2010-02-05        42.31       2.572  211.096358         8.106   
3      1  2010-02-05        42.31       2.572  211.096358         8.106   
4      1  2010-02-05        42.31       2.572  211.096358         8.106   

   IsHoliday  Markdowns    Size  Dept  Weekly_Sales  
0      False        0.0  151315     1      24924.50  
1      False        0.0  151315     2      50605.27  
2      False        0.0  151315     3      13740.12  
3      False        0.0  151315     4      39954.04  
4      False        0.0  151315     5      32229.38

#### fix date into two columns one for month and other for year


In [9]:
from pandas import DatetimeIndex
df = Merged_Train_Features_Store_Df
df.Date = pd.to_datetime(df.Date)
    

In [10]:
df['Year'] = DatetimeIndex(df['Date']).year
df['Month']= DatetimeIndex(df['Date']).month
df['Day'] = DatetimeIndex(df['Date']).day
df = df.drop(columns=['Date'])
df[0::1000]

Store  Temperature  Fuel_Price         CPI  Unemployment  IsHoliday  \
0             1        42.31       2.572  211.096358         8.106      False   
1000          1        42.31       2.572  211.096358         8.106      False   
2000          1        42.31       2.572  211.096358         8.106      False   
3000          2        40.19       2.572  210.752605         8.324      False   
4000          2        40.19       2.572  210.752605         8.324      False   
5000          2        40.19       2.572  210.752605         8.324      False   
6000          3        45.71       2.572  214.424881         7.368      False   
7000          3        45.71       2.572  214.424881         7.368      False   
8000          3        45.71       2.572  214.424881         7.368      False   
9000          4        43.76       2.598  126.442065         8.623      False   
10000         4        43.76       2.598  126.442065         8.623      False   
11000         4        43.76       2.598  126.442065         8.623      False   
12000         5        39.70       2.572  211.653972         6.566      False   
13000         5        39.70       2.572  211.653972         6.566      False   
14000         5        39.70       2.572  211.653972         6.566      False   
15000         6        40.43       2.572  212.622352         7.259      False   
16000         6        40.43       2.572  212.622352         7.259      False   
17000         6        40.43       2.572  212.622352         7.259      False   
18000         7        10.53       2.580  189.381697         9.014      False   
19000         7        10.53       2.580  189.381697         9.014      False   
20000         7        10.53       2.580  189.381697         9.014      False   
21000         8        34.14       2.572  214.471451         6.299      False   
22000         8        34.14       2.572  214.471451         6.299      False   
23000         8        34.14       2.572  214.471451         6.299      False   
24000         9        38.01       2.572  214.655459         6.415      False   
25000         9        38.01       2.572  214.655459         6.415      False   
26000         9        38.01       2.572  214.655459         6.415      False   
27000        10        54.34       2.962  126.442065         9.765      False   
28000        10        54.34       2.962  126.442065         9.765      False   
29000        10        54.34       2.962  126.442065         9.765      False   
...         ...          ...         ...         ...           ...        ...   
18941000     35        58.99       3.882  142.762411         8.665      False   
18942000     36        74.39       3.494  222.113657         6.228      False   
18943000     36        74.39       3.494  222.113657         6.228      False   
18944000     36        74.39       3.494  222.113657         6.228      False   
18945000     37        74.39       3.506  222.113657         6.228      False   
18946000     37        74.39       3.506  222.113657         6.228      False   
18947000     38        65.95       4.301  131.193097        10.199      False   
18948000     38        65.95       4.301  131.193097        10.199      False   
18949000     38        65.95       4.301  131.193097        10.199      False   
18950000     39        72.90       3.506  222.113657         6.228      False   
18951000     39        72.90       3.506  222.113657         6.228      False   
18952000     39        72.90       3.506  222.113657         6.228      False   
18953000     40        49.65       3.917  138.728161         4.145      False   
18954000     40        49.65       3.917  138.728161         4.145      False   
18955000     40        49.65       3.917  138.728161         4.145      False   
18956000     41        41.80       3.686  199.219532         6.195      False   
18957000     41        41.80       3.686  199.219532         6.195      False   
18958000     41        41.80       3.686 

### Split the merged data set into X_train and y_train, use test dataset for X_test

In [11]:
#split with method

y =df['Weekly_Sales']
X = df.drop(columns=['Weekly_Sales'])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, 
                                                    random_state=2)

In [12]:
#split into train and test

# X_train = Merged_Train_Features_Store_Df[['Store','Date','Temperature','Fuel_Price','CPI','Unemployment','IsHoliday','Markdowns','Size','Dept']]


# X_train.head()

# y_train = Merged_Train_Features_Store_Df[['Weekly_Sales']]

# X_test = walmart_test_df.drop(columns=['Date'])

In [13]:
#Linear Regression

linear_reg = LinearRegression()
linear_reg.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [14]:
print("Intercept: ", linear_reg.intercept_)

print("Coefficient: " , linear_reg.coef_)

Intercept:  510101.1788285926
Coefficient:  [-3.42340540e-01 -1.82609942e+01  1.25629623e+02  2.35953141e+00
  2.96616567e+01  4.61274125e+02  1.28013425e-02 -1.05286129e-03
  1.10048300e+02 -2.48643105e+02  2.07380750e+02 -1.55636958e+01]


### predictive model

$$y = 510101.179 - 0.342 \times Store - 18.26 \times Temperature + 125.630 \times Fuel Price + 2.359 \times CPI + 29.661 \times Unemployment + 461.274 \times IsHoliday + 0.0128 \times Markdowns - 0.0010 \times Size + 110.048 \times Department - 248.643 \times Year + 207.381 \times Month - 15.564 \times Day $$

### Testing and prediction

In [15]:
y_pred = linear_reg.predict(X_test)
print(y_pred)

[15543.17842783 21221.08744366 15849.74566892 ... 15035.94276931
 14853.15899795 16203.75431315]


### RMSE

In [16]:
from sklearn import metrics
mse = metrics.mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)

In [17]:
print(rmse)

22462.71900529855


## Cross Validation

In [23]:
logi_reg = LogisticRegression()
acc_list = cross_val_score(logi_reg, X, y, cv =10)
print(acc_list)

ValueError: Unknown label type: 'continuous'

In [ ]:
acc_cv = acc_list.mean()
print(acc_cv)